### ISSUE TITLE
Astropy Tables assignment to .loc

### ISSUE THREAD

**Issue by karlglazebrook** [astropy/astropy#13530]
### Description

I am not sure if this is a bug or a feature request, or a misunderstanding of the docs on my part, it concerns assigning to Tables using .loc

### Here is the code

```
t = Table([(1, 2, 3, 4), (10, 1, 9, 9)], names=('a', 'b'), dtype=['i8', 'i8'])
t.add_index('a')
print(t)
# I expect this to change the value of first row column 'b'
t.loc[1]['b'] = 42
# And it does
print(t)
# But this does not change the values of the next two rows, and I expect it to
t.loc[[2,3]]['b'] = [43,44]
print(t)

```
### Here is the output

 ```
a   b 
--- ---
  1  10
  2   1
  3   9
  4   9
 a   b 
--- ---
  1  42
  2   1
  3   9
  4   9
 a   b 
--- ---
  1  42
  2   1
  3   9
  4   9

```

### System Details

Darwin-18.7.0-x86_64-i386-64bit
Python 3.7.6 (default, Jan  8 2020, 13:42:34) 
[Clang 4.0.1 (tags/RELEASE_401/final)]
Numpy 1.21.6
pyerfa 2.0.0.1
astropy 4.0
Scipy 1.7.3
Matplotlib 3.5.2

**Comment by taldcroft** [comment]
@karlglazebrook - this behavior is expected and is not likely something we want to change. 

tl;dr: what you want to do is this:
```
t['b'][t.loc_indices[[2,3]]] = [43, 44]
```

### What's going on in your original example

You can see the same without using `.loc`:
```
>>> t = Table([(1, 2, 3, 4), (10, 1, 9, 9)], names=('a', 'b'), dtype=['i8', 'i8'])
>>> t[[1,2]]['b'] = 100
>>> t
<Table length=4>
  a     b  
int64 int64
----- -----
    1    10
    2     1
    3     9
    4     9
```
It should be more clear if you break down `t[[1,2]]['b'] = 100` into two steps:
```
t2 = t[[1,2]]  # this is a new table with a *copy* of the data
t2['b'] = 100
```
Note that when dealing with a single array or column, something like `t['b'][[1,2]] = 100` is calling `__setitem__` on `t['b']`, but `t[[1,2]]['b']` is making a new intermediate table and then setting the `b` column of that intermediate table.

**Comment by mhvk** [comment]
To add to @taldcroft's comment: numpy arrays have the same behaviour - mostly because in general it is not possible to return a view of the data if one uses a list of indices (unlike for a slice, which does work):
```
In [2]: a = np.array([(1, 2), (3, 3)], dtype='i,i')

In [3]: a
Out[3]: array([(1, 2), (3, 3)], dtype=[('f0', '<i4'), ('f1', '<i4')])

In [4]: a[[0,1]]
Out[4]: array([(1, 2), (3, 3)], dtype=[('f0', '<i4'), ('f1', '<i4')])

In [5]: a[[0,1]]['f0'] = 10

In [6]: a
Out[6]: array([(1, 2), (3, 3)], dtype=[('f0', '<i4'), ('f1', '<i4')])

In [7]: a[:]['f0'] = 10

In [8]: a
Out[8]: array([(10, 2), (10, 3)], dtype=[('f0', '<i4'), ('f1', '<i4')])
```

**Comment by karlglazebrook** [comment]
@mhvk Are you sure about this?

```
With your example

In [1]: a = np.array([(1, 2), (3, 3)], dtype='i,i')

In [2]: a
Out[2]: array([(1, 2), (3, 3)], dtype=[('f0', '<i4'), ('f1', '<i4')])

In [3]: a[[0,1]] = 10

In [4]: a
Out[4]: array([(10, 10), (10, 10)], dtype=[('f0', '<i4'), ('f1', '<i4')])
```

The only difference is I did not include the 'f0' part, which I do not understand.

**Comment by karlglazebrook** [comment]
@taldcroft Regarding

`Note that when dealing with a single array or column, something like t['b'][[1,2]] = 100 is calling __setitem__ on t['b'], but t[[1,2]]['b'] is making a new intermediate table and then setting the b column of that intermediate table.
`


I understand your point, but I would expect  both `t[[1,2]]` and `t.loc[[1,2]] `to return new *views* in to `t `which are settable. Why is my expectation wrong?

I would also expect` t[[1,2]]['b'] =100` and` t['b'][[1,2]] =100 `to do the same thing

**Comment by taldcroft** [comment]
@karlglazebrook - you are missing the distinction between *setting* with a list (`a[[0,1]] = 10`) and *indexing* with a list (below):
```
In [15]: a = np.array([(1, 2), (3, 3)], dtype="i,i")

In [17]: b = a[[0, 1]]  # This makes a copy. That's that way numpy and astropy Table works.

In [18]: b[()] = 10  # Set every element of b to 10

In [19]: b
Out[19]: array([(10, 10), (10, 10)], dtype=[('f0', '<i4'), ('f1', '<i4')])

In [20]: a
Out[20]: array([(1, 2), (3, 3)], dtype=[('f0', '<i4'), ('f1', '<i4')])
```
The fact that `b = a[[0, 1]]` is a copy and not a view is related to implementation. If this returned a view then that view object would need to keep a copy of the original slice list within `b` in order to maintain the view into `a`. The original developer of numpy (and its predecessors) decided the performance and memory hit was not worth it. 

All I can say is that you need to adjust your expectations at this point. :smile:

**Comment by karlglazebrook** [comment]
Happy to adjust. :-) However if if I do

```
a[[0, 1]]=10
a
array([(10, 10), (10, 10)], dtype=[('f0', '<i4'), ('f1', '<i4')])
```

So it is a view. This is why I expect `t[[1,2]]['b'] =100` and `t['b'][[1,2]] =100` to do the same thing? I am not sure I understand the logic as to why it is creating an intermediate in one case but not the other.

**Comment by taldcroft** [comment]
Under the hood, `a[[0,1]] = 10` is equivalent to `a.__setitem__([0, 1], 10)`. Python is calling a method on the original `a` object to set items. You might have imagined that this is like `a.__getitem__([0, 1]) = 10`, but that's not how Python works.

For the second case `t[[1,2]]['b'] = 100` is equivalent to `t.__getitem__([1,2]).__setitem__('b', 100)`. That's where the copy comes in.

The third case `t['b'][[1,2]] =100` is `t.__getitem__('b').__setitem__([1,2], 100)`. Since `t.__getitem__('b')` returns the original column `b` then setting it works like you expect.

**Comment by karlglazebrook** [comment]
So you are saying that ` t.__getitem__('b')` returns a view in to a column, which is why it is settable as an lvalue, but` t.__getitem__([1,2])` returns a copy which is not.

But ` t.__getitem__([1,2])` corresponds to two rows of `t`, so why can't that be a view too rather than a new table? That is what I would have logically expected coming from other languages which do indeed do this. Why are columns privileged over rows here?

I see if I do `t[1:2]['b'] =100`  that does work, so is it simply that this is an arbitrary underlying numpy limitation not to be able to handle a general list of rows as a view - as opposed to a slice?  

 I note that even in numpy it behaves as expected if it is **not** a structured array. So a limitation of structured arrays?

**Comment by mhvk** [comment]
@karlglazebrook - yes, it is a limitation of numpy, since in general an index with a list can be anything: it is not possible to  return a view of `__getitem__([1, 2, 4])` (because any array needs constant strides along its dimensions). Note that `__setitem__([1, 2, 4], something)` does work because that follows a different path. But as @taldcroft noted, when you are doing something like `a[i1][i2] = b`, you get a.__getitem__(i1).__setitem__(i2, b)` and things no longer work if `[i1]` does not return a view.

**Comment by taldcroft** [comment]
> Why are columns privileged over rows here?

Columns are not privileged over rows, they behave the same. `t['a']` is a view of column `a` and `t[1]` is a view of row 1. `t[['a', 'b']]` is a copy of columns `a` and `b`, and `t[[1, 2]]` is a copy of rows 1 and 2. Lastly `t.columns[0:2]` is a view slice of the first two columns `a` and `b` just like `t[0:2]` is a view slice of rows 0 and 1.

There are some asymmetries related to syntax but in general the behaviors are consistent within astropy Table and with how numpy and pandas work.

**Comment by karlglazebrook** [comment]
There are indeed some asymmetries.

```
t = Table([(1, 2, 3, 4), (10, 1, 9, 9)], names=('a', 'b'), dtype=['i8', 'i8'])
t[[1, 2]]=42 # Works and modifies t
t[[1, 2]]['b']=42 # Works but does NOT modify  t
t['b'][[1, 2]]=42 # Works and modifies t
t['b'] = 42 # Works and modifies t
t[['a', 'b']] = 42 # Returns an error
```

If I understand the discussion correctly it is due to limitations of python and numpy. Viz `t.__getitem__('b')`  is different to `t.__getitem__([1, 2,]) ` when further qualified because of limitations arounds indexing being views.

Hope I have got that right! I will learn to live with it and thanks all for help. I still kind of think that in principle they should all work :-)
